### PR TITLE
BUG: encoding of categoricals in hdf serialization

### LINKED ISSUE #10366: Irregular errors when reading certain categorical strings from hdf
It seems that there is something bad happening when we use certain strings with special characters AND the empty string with categoricals:

```python
    # -*- coding: latin-1 -*-
    import pandas
    import os

    examples = [
            pandas.Series(['EÉ, 17', '', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['EÉ, 17', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['EE, 17', '', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['øü', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['Aøü', '', 'a', 'b', 'c'], dtype='category'),
            pandas.Series(['EÉ, 17', 'øü', 'a', 'b', 'c'], dtype='category')
            ]

    def test_hdf(s):
        f = 'testhdf.h5'
        if os.path.exists(f):
            os.remove(f)
        s.to_hdf(f, 'data', format='table')
        return pandas.read_hdf(f, 'data')

    for i, s in enumerate(examples):
        flag = True
        e = ''
        try:
            test_hdf(s)
        except Exception as ex:
            e = ex
            flag = False
        print('%d: %s\t%s\t%s' % (i, 'pass' if flag else 'fail', s.tolist(), e))
```
Results in:

```
    0: fail ['EÉ, 17', '', 'a', 'b', 'c']   Categorical categories must be unique
    1: pass ['EÉ, 17', 'a', 'b', 'c']
    2: pass ['', 'a', 'b', 'c']
    3: pass ['EE, 17', '', 'a', 'b', 'c']
    4: pass ['øü', 'a', 'b', 'c']
    5: fail ['Aøü', '', 'a', 'b', 'c']  Categorical categories must be unique
    6: pass ['EÉ, 17', 'øü', 'a', 'b', 'c']
```

Not sure if I am using this incorrectly or if this is actually a corner case.

### PR DESCRIPTION
**Pull request by jreback**
closes #10366 
replaces #10454

### CODE REVIEW

**Comment by jreback** [comment]
cc @cottrell 

I changed this a bit to explicity check for bytes, and doced this a bit.

Seems ok to me, but can you check for

File: doc/source/whatsnew/v0.17.0.txt [hunk 1]
```diff
@@ -765,3 +765,4 @@ Bug Fixes
 - Bug in ``read_msgpack`` where encoding is not respected (:issue:`10580`)
 - Bug preventing access to the first index when using ``iloc`` with a list containing the appropriate negative integer (:issue:`10547`, :issue:`10779`)
 - Bug in ``TimedeltaIndex`` formatter causing error while trying to save ``DataFrame`` with ``TimedeltaIndex`` using ``to_csv`` (:issue:`10833`)
+- Bug in ``Categorical`` hdf serialiation in presence of alternate encodings. (:issue:`10366`)
```
Comment (kawochen) [inline doc/source/whatsnew/v0.17.0.txt]: Typo in serialization

**Comment by cottrell** [comment]
nosetests -v pandas.io.tests.test_pytables passed on my local python 3.4.3 build. I see other errors popping up as in the travis tests.

**Comment by jreback** [comment]
thanks @cottrell

**Pull request merged**